teps/0046-triggertemplate-parameter-escaping.md

```diff
@@ -0,0 +1,271 @@
+it does feel somewhat odd that the template insertion has to know whether or not
+the data should be escaped, clearly you could use a parameter in multiple places
+within a document, and in some of those places, escaping might be required or
+more appropriate than others, but it would be configured at the TriggerTemplate
+level, and not at the insertion level.
+
+### Risks and Mitigations
+
+We already have escaping, which can now be disabled, and the proposal is really
+just a fine-tuning of the functionality to allow users to control which strings
+they need to escape.
+
+### User Experience (optional)
+
+### Performance (optional)
+
+There's unlikely to be a significant performance hit from this approach, as the
+functionality already exists.
+
+## Design Details
+
+This would extend the [current](https://github.com/tektoncd/triggers/blob/master/pkg/apis/triggers/v1alpha1/param.go#L5-L15) `ParamSpec` struct.
+
+```
+type ParamSpec struct {
+    < existing fields omitted >>
+
+	// Escape indicates whether or not the value should be escaped as JSON
+	// before inserting into the templated bodies.
+	Escape bool `json:"escape,omitempty"`
+}
+```
+
+This would translate into the YAML as an addition to the declaration:
+
+```yaml
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: TriggerTemplate
+metadata:
+  name: escaped-param
+spec:
+  params:
+  - name: title
+    description: The title from the incoming body
+    escape: true
+```
+
+And in the code, we'd apply the current "old escape" logic, i.e. replacing `"`
+with `\"`, only on the parameters marked as requiring to be escaped, but there's
+definitely scope for alternative escaping.
+
+## Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+## Design Evaluation
+
+
+## Drawbacks
+
+I'm not aware of any, it's just altering existing behaviour to make it more
+flexible.
+
+It should allow dropping of the TriggerTemplate-wide annotation to indicate that
+all parameters should be escaped.
+
+## Alternatives
+
+When the fix for [#777](https://github.com/tektoncd/triggers/pull/823) was
+landed, the annotation-based approach was applied, this allowed existing
+templates to continue to work, but fixed the issue with over-escaping values.
+
+The annotation approach was considered temporary at that point, but we could
+consider this permanent.
+
+Another would be to change the escaping to be some sort of strategy based
```

Comment:
Another option I was discussing with @wlynch was having the escaping done at the TriggerBinding level not at the TriggerTemplate level. Is there ever a case where we extract a value that contains an escaped quote but then we'd sometimes want to escape it but not always?

Reply:
I mulled this over a lot, both when implementing tektoncd/triggers#823 and writing this up.

I had thought about applying it to TriggerBindings something like this:

```yaml
apiVersion: triggers.tekton.dev/v1alpha1
kind: TriggerBinding
metadata:
  name: pipeline-binding
spec:
  params:
  - name: gitrevision
    value: $(body.head_commit.id)
    escapeQuotes: true
```

And when I was writing this:

> Ultimately, this is about escaping external data being received into the system,
> it does feel somewhat odd that the template insertion has to know whether or not
> the data should be escaped...

I reasoned that the entire purpose of triggers is to generate templated resources from incoming data, and so differentiation on whether or not "tainted" values should be escaped could be deemed to be moot, because all data is "tainted", which got me to the view that perhaps the issue is with the way it's escaped (string replace quotes), and not with the escaping itself, which is what got me to different strategies for escaping data.

This then got me back to, yes, the TriggerTemplate is probably the best place for this, because you're inserting the value into the body, and it has the context of where the value is being inserted, and so, I reconsidered the `%(param.name)` approach.

Another approach entirely, which was also suggested by @chmouel (and is probably not in scope for this particular change) would be to apply Jinja (or other another templating languge) to the resources?

Reply:
> Another approach entirely, which was also suggested by @chmouel (and is probably not in scope for this particular change) would be to apply Jinja (or other another templating languge) to the resources?

Yeah using an actual templating engine has come up before. I think in the past we decided that we wanted to keep things as simple as possible so just simple variable interpolation would be sufficient. Though as we found out there are a few edge cases (like newlines, quotes) with such an approach.

> This then got me back to, yes, the TriggerTemplate is probably the best place for this, because you're inserting the value into the body, and it has the context of where the value is being inserted, and so, I reconsidered the %(param.name) approach.

I think I agree but the way I was thinking about adding it to Bindings instead is that bindings should only capture "valid" values.  For our use case, any value that is extracted has to be usable in a JSON document. If we do this at the binding level, we only need to do it once per binding. Otherwise, we'd have to do it for each trigger template that uses that binding which sounds less user friendly.

Reply:
In Tekton based CI we use bindings to create an interface for CI jobs - we provide a fixed set of parameters, so that a user writing a CI job does not have to worry about events, interceptors or bindings. With this in mind, the person writing the bindings does not have the context to decide whether a parameter should be escaped or not, because they don't actually know how the parameter is going to be used.